File: ubf_app/lib/core/utils/payment_state.dart
import 'package:mana/l10n/app_localizations.dart';

/// 입금 상태 — 낼 돈과 낸 돈에서 하나로 정한다.
///
/// **서버의 payment_state.js 와 같은 규칙이다.** 두 곳에 두는 것이 마음에
/// 들지는 않지만, 표는 이미 받아 둔 행으로 그 자리에서 다시 그려야 하고
/// 서버에 다시 묻지 않는다. 규칙이 바뀌면 양쪽을 함께 고쳐야 한다 —
/// 그래서 판정 자체는 이 한 함수에만 둔다.
enum PayState { unpaid, partial, paid, pending }

PayState payStateOf({num? due, num? paid, String? status}) {
  final owed = (due ?? 0) > 0 ? (due ?? 0) : 0;
  // 확인된 것만 낸 것으로 센다. 거절은 낸 것이 아니다.
  final got = (status == 'confirmed' && (paid ?? 0) > 0) ? (paid ?? 0) : 0;

  // 확인 전에는 금액과 상관없이 대기 — 확인하지 않은 돈을 세면 장부가
  // 실제보다 커진다.
  if (status == 'pending') return PayState.pending;
  if (owed == 0) return PayState.paid;
  if (got <= 0) return PayState.unpaid;
  if (got < owed) return PayState.partial;
  return PayState.paid;
}

String payStateLabel(AppLocalizations l10n, PayState s) => switch (s) {
  PayState.unpaid => l10n.payUnpaid,
  PayState.partial => l10n.payPartial,
  PayState.paid => l10n.payPaid,
  PayState.pending => l10n.payPending,
};

File: server/src/services/payment_state.js
// 입금 상태 — 낼 돈과 낸 돈에서 하나로 정한다.
//
// DB 도 HTTP 도 쓰지 않는다. test/payment_state.test.js 로 그대로 확인한다.
//
// **상태를 따로 저장하지 않는다.** 금액과 상태를 둘 다 적어 두면 언젠가
// 어긋난다 — 20만원 중 5만원을 받았는데 상태는 '완납' 인 줄이 생긴다.
// 금액에서 매번 계산한다.

export const PAYMENT_STATES = ['unpaid', 'partial', 'paid', 'pending'];

/// 얼마를 냈는지 세는 상태. 거절(rejected)은 낸 것이 아니다.
function countsAsPaid(status) {
  return status === 'confirmed';
}

/// 넷 중 하나.
///
///   pending — 냈다고 알려 왔지만 담당자가 아직 확인하지 않았다
///   unpaid  — 확인된 금액이 없다
///   partial — 일부만 확인됐다
///   paid    — 낼 돈을 다 채웠다
///
/// **확인 전에는 금액과 상관없이 '대기'** 다. 확인하지 않은 돈을 받은 것으로
/// 세면 장부가 실제보다 커진다.
///
/// 낼 돈이 0 이면(참가비를 안 정한 수양회, 전액 면제) 낼 것이 없으므로
/// 완납으로 본다 — '미납' 이라고 하면 담당자가 받으러 다닌다.
export function paymentState({ due, paid, status } = {}) {
  const d = Number(due);
  const p = Number(paid);
  const owed = Number.isFinite(d) && d > 0 ? d : 0;
  const got = Number.isFinite(p) && p > 0 && countsAsPaid(status) ? p : 0;

  if (status === 'pending') return 'pending';
  if (owed === 0) return 'paid';
  if (got <= 0) return 'unpaid';
  if (got < owed) return 'partial';
  return 'paid';
}

/// 아직 받을 돈. 화면이 "얼마 남았나" 를 묻는다.
export function remaining({ due, paid, status } = {}) {
  const d = Number(due);
  const p = Number(paid);
  const owed = Number.isFinite(d) && d > 0 ? d : 0;
  const got = Number.isFinite(p) && p > 0 && countsAsPaid(status) ? p : 0;
  return Math.max(0, owed - got);
}
